client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +146,154 @@ public void start(Listener listener) {
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                if (this.enableDiscovery) {
+                    // Make an RPC call to the bootstrapped controller servers to fetch all active controllers.
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server hostnames to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
+                        final InetSocketAddress socketAddress = new InetSocketAddress(address.getHostString(),
+                                address.getPort());
+                        if (!socketAddress.isUnresolved()) {
+                            resolvedAddresses.add(socketAddress);
+                        }
+                    });
+                    serverInfoGroup = ResolvedServerInfoGroup.builder().addAll(resolvedAddresses.stream()
+                            .map(ResolvedServerInfo::new)
+                            .collect(Collectors.toList()))
+                            .build();
+                }
 
                 // Update gRPC load balancer with the new set of server addresses.
                 log.info("Updating client with controllers: {}", serverInfoGroup);
                 this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
-            } catch (StatusRuntimeException e) {
-                log.warn("Failed to fetch controller addresses - {}", e);
-                this.resolverUpdater.onError(e.getStatus());
+
+                // We have found at least one controller endpoint. Repeat discovery after the regular schedule.
+                nextScheduleTimeMS = REFRESH_INTERVAL_MS;
+            } catch (Throwable e) {
+                // Catching all exceptions here since this method should never exit without rescheduling the discovery.
+                if (e instanceof StatusRuntimeException) {
+                    this.resolverUpdater.onError(((StatusRuntimeException) e).getStatus());
```

Comment:
> Access to the var here cannot contend with the assignment

That is simply not true. In the Java memory model things are allowed to be arranged if there is no memory barrier. For example:
Thread 1 calls start() is assigns the listener.
Thread 2 calls refresh() because there is an @synchronize block everything in thread 2 happens after the call to start(). But at the end of refresh it sets up a call to getControllers() to run in a background thread.
The background thread starts and executes getControllers(), there is no synchronization encountered so it is free to that CPU core's cache of the variables rather than accessing main ram, it could so happen that this core has a cached version of the object before the assignment of resolverUpdater from before it was assigned in start().
So even though the external caller didn't do anything other than use this class as the interface dictated this line will thrown an NPE and it will take a long time to figure out why.

The threading rules that we put together in our contributing guidelines exist for a reason. They are a sane way to make sure everything is threadsafe. Please follow them. Because anytime you don't, I'll bet money that there is a race condition you haven't thought about, like the one here.